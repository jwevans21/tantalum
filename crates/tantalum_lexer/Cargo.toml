[package]
name = "tantalum_lexer"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}
pretty_assertions = { workspace = true }
rstest = { workspace = true }


[lints]
workspace = true
