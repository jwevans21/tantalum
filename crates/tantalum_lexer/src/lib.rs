//! # Tantalum Lexer
//!
//! This crate provides a lexer for the Tantalum language.

use tantalum_source::{SourceFile, SourceLocation, SourceSpan};
use tantalum_syntax::{SyntaxKind, SyntaxToken};

/// A lexer for the Tantalum language.
#[derive(Debug, Clone, PartialEq, Eq, Hash)]
pub struct Lexer<'file> {
    /// The source file that the lexer is currently lexing.
    file: &'file SourceFile,
    /// The current location in the source file.
    location: SourceLocation,
}

impl<'file> Lexer<'file> {
    /// Create a new lexer from a source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let lexer = Lexer::new(file);
    /// ```
    #[must_use]
    pub fn new(file: &'file SourceFile) -> Self {
        Self {
            file,
            location: file.location_start(),
        }
    }

    /// Get the source file that the lexer is currently lexing.
    #[must_use]
    pub fn source(&self) -> &'file SourceFile {
        self.file
    }

    /// Get the current location in the source file.
    #[must_use]
    pub fn location(&self) -> SourceLocation {
        self.location
    }

    /// Get the character at a location in the source file.
    fn at_location(&self, location: SourceLocation) -> Option<char> {
        self.file.read_char(location)
    }

    /// Extract the next token from the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.next_token().expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    /// ```
    #[must_use]
    pub fn next_token(&mut self) -> Option<SyntaxToken> {
        let start = self.location;

        let (kind, end) = self.characters_to_token(start)?;

        self.location = end;

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Peek at the next token in the source file.
    ///
    /// ```rust
    /// use tantalum_lexer::Lexer;
    /// use tantalum_source::SourceFileCollection;
    /// use tantalum_syntax::SyntaxKind;
    ///
    /// let mut files = SourceFileCollection::new();
    ///
    /// let file_id = files.add_file("file.tan".to_string(), "fn main() {}".to_string());
    ///
    /// let file = files.get_file(file_id).expect("file should exist since it was just added");
    ///
    /// let mut lexer = Lexer::new(file);
    ///
    /// let token = lexer.peek_token(1).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Fn);
    ///
    /// let token = lexer.peek_token(2).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Whitespace);
    ///
    /// let token = lexer.peek_token(3).expect("token should exist since the file is not empty");
    ///
    /// assert_eq!(token.kind(), SyntaxKind::Identifier);
    /// ```
    #[must_use]
    pub fn peek_token(&self, mut n: usize) -> Option<SyntaxToken> {
        let mut start = self.location;

        let (kind, end) = loop {
            n -= 1;
            let (kind, end) = self.characters_to_token(start)?;

            if n == 0 {
                break (kind, end);
            }

            start = end;
        };

        Some(SyntaxToken::new(
            kind,
            SourceSpan::new(self.file.id(), start.offset(), end.offset()),
        ))
    }

    /// Convert a character to a [`SyntaxKind`].
    ///
    /// [`SyntaxKind`]: tantalum_syntax::SyntaxKind
    fn characters_to_token(&self, start: SourceLocation) -> Option<(SyntaxKind, SourceLocation)> {
        let mut current = start;

        let c = self.at_location(current)?;
        current = current.next_by(c.len_utf8());

        let token = match c {
            '(' => SyntaxKind::LParen,
            ')' => SyntaxKind::RParen,
            '{' => SyntaxKind::LBrace,
            '}' => SyntaxKind::RBrace,
            '[' => SyntaxKind::LBracket,
            ']' => SyntaxKind::RBracket,
            '<' => SyntaxKind::LAngle,
            '>' => SyntaxKind::RAngle,
            ',' => SyntaxKind::Comma,
            ':' => SyntaxKind::Colon,
            ';' => SyntaxKind::Semicolon,
            '.' => SyntaxKind::Dot,
            '&' => SyntaxKind::Ampersand,
            '|' => SyntaxKind::Pipe,
            '^' => SyntaxKind::Caret,
            '+' => SyntaxKind::Plus,
            '-' => SyntaxKind::Minus,
            '*' => SyntaxKind::Star,
            '/' => SyntaxKind::Slash,
            '\\' => SyntaxKind::Backslash,
            '%' => SyntaxKind::Percent,
            '=' => SyntaxKind::Equals,
            '!' => SyntaxKind::Bang,
            '~' => SyntaxKind::Tilde,
            '?' => SyntaxKind::Question,
            '@' => SyntaxKind::At,
            '0'..='9' => SyntaxKind::Digit,
            '_' | 'a'..='z' | 'A'..='Z' => {
                while let Some(c) = self.at_location(current) {
                    match c {
                        '_' | 'a'..='z' | 'A'..='Z' | '0'..='9' => {
                            current = current.next_by(c.len_utf8());
                        }
                        _ => break,
                    }
                }

                match self.file.read_span(SourceSpan::new(
                    self.file.id(),
                    self.location.offset(),
                    current.offset(),
                ))? {
                    "fn" => SyntaxKind::Fn,
                    "let" => SyntaxKind::Let,
                    "return" => SyntaxKind::Return,
                    "if" => SyntaxKind::If,
                    "else" => SyntaxKind::Else,
                    "true" => SyntaxKind::True,
                    "false" => SyntaxKind::False,
                    _ => SyntaxKind::Identifier,
                }
            }
            c if c.is_whitespace() => SyntaxKind::Whitespace,
            _ => SyntaxKind::Other,
        };

        Some((token, current))
    }
}

#[test]
fn keywords() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "fn let return if else true false".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 7);

    assert_eq!(tokens[0].kind(), SyntaxKind::Fn);
    assert_eq!(tokens[1].kind(), SyntaxKind::Let);
    assert_eq!(tokens[2].kind(), SyntaxKind::Return);
    assert_eq!(tokens[3].kind(), SyntaxKind::If);
    assert_eq!(tokens[4].kind(), SyntaxKind::Else);
    assert_eq!(tokens[5].kind(), SyntaxKind::True);
    assert_eq!(tokens[6].kind(), SyntaxKind::False);
}

#[test]
fn identifiers() {
    use crate::Lexer;
    use tantalum_source::SourceFileCollection;
    use tantalum_syntax::SyntaxKind;

    let mut files = SourceFileCollection::new();

    let file_id = files.add_file(
        "file.tan".to_string(),
        "a A _ a1 A1 _1 _a _A __AA01".to_string(),
    );

    let file = files
        .get_file(file_id)
        .expect("file should exist since it was just added");

    let mut lexer = Lexer::new(file);

    let mut tokens = Vec::new();
    while let Some(token) = lexer.next_token() {
        if token.kind() == SyntaxKind::Whitespace {
            continue;
        }
        tokens.push(token);
    }

    assert_eq!(tokens.len(), 9);

    assert_eq!(tokens[0].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[1].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[2].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[3].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[4].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[5].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[6].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[7].kind(), SyntaxKind::Identifier);
    assert_eq!(tokens[8].kind(), SyntaxKind::Identifier);
}
