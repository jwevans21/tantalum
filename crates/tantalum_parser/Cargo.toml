[package]
name = "tantalum_parser"
version = "0.1.0"
authors.workspace = true
edition.workspace = true
license.workspace = true
readme.workspace = true
repository.workspace = true
publish.workspace = true

[dependencies]
serde = { workspace = true, optional = true }

tantalum_ast = { path = "../tantalum_ast" }
tantalum_lexer = { path = "../tantalum_lexer" }
tantalum_span = { path = "../tantalum_span" }

[features]
default = ["serde"]
serde = ["dep:serde"]

[dev-dependencies]
insta = { workspace = true}

[lints]
workspace = true
