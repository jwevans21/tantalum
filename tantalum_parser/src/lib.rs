use error::ParseError;
use tantalum_ast::TopLevelExpression;
use tantalum_lexer::{token::Token, token_kind::TokenKind, Lexer};
use tantalum_span::Span;

pub mod error;

mod expressions;
mod statements;
mod types;
mod top_level;

#[cfg(test)]
mod tests;

#[derive(Debug, Clone, PartialEq, Eq)]
#[non_exhaustive]
pub struct Parser<'file_name, 'source> {
    // lexer: Lexer<'file_name, 'source>,
    source: &'source str,
    file_name: &'file_name str,
    tokens: Vec<Token<'file_name, 'source>>,
    eof: Span<'file_name>,
    position: usize,
}

impl<'file_name, 'source> Parser<'file_name, 'source> {
    #[must_use]
    #[inline]
    pub fn new(mut lexer: Lexer<'file_name, 'source>) -> Self {
        Self {
            source: lexer.source(),
            file_name: lexer.file_name(),
            tokens: lexer.by_ref().collect(),
            eof: lexer.span(),
            position: 0,
        }
    }

    /// Parse the entire source file.
    /// 
    /// # Errors
    /// 
    /// Returns an error if the parser encounters an unexpected token or the end of the file.
    pub fn parse(
        &mut self,
    ) -> Result<Vec<TopLevelExpression<'file_name, 'source>>, error::ParseError<'file_name, 'source>>
    {
        let mut top_levels = Vec::new();

        while !self.is_eof() {
            top_levels.push(self.parse_top_level()?);
        }

        Ok(top_levels)
    }

    fn is_eof(&self) -> bool {
        self.position >= self.tokens.len()
    }

    fn is_at(&self, kind: TokenKind) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position)
            .filter(|token| token.kind() == kind)
            .copied()
    }

    fn advance_if(&mut self, kind: TokenKind) -> Option<Token<'file_name, 'source>> {
        if let Some(token) = self.is_at(kind) {
            self.position += 1;
            Some(token)
        } else {
            None
        }
    }

    fn expect(
        &mut self,
        kind: TokenKind,
    ) -> Result<Token<'file_name, 'source>, error::ParseError<'file_name, 'source>> {
        let Some(token) = self.tokens.get(self.position) else {
            return Err(error::ParseError::unexpected_eof(self.source, self.eof));
        };

        if token.kind() == kind {
            self.position += 1;
            Ok(*token)
        } else {
            Err(error::ParseError::unexpected_token_set(
                self.source,
                token.span(),
                token.kind(),
                &[kind],
            ))
        }
    }

    #[allow(unused)]
    fn next_is(&self, kind: TokenKind) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position + 1)
            .filter(|token| token.kind() == kind)
            .copied()
    }

    fn is_at_any<'a>(&self, set: &'a [TokenKind]) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position)
            .filter(|token| set.contains(&token.kind()))
            .copied()
    }

    fn advance_if_any<'a>(&mut self, set: &'a [TokenKind]) -> Option<Token<'file_name, 'source>> {
        if let Some(token) = self.is_at_any(set) {
            self.position += 1;
            Some(token)
        } else {
            None
        }
    }

    fn expect_any<'a>(
        &mut self,
        set: &'a [TokenKind],
    ) -> Result<Token<'file_name, 'source>, error::ParseError<'file_name, 'source>> {
        let Some(token) = self.tokens.get(self.position) else {
            return Err(error::ParseError::unexpected_eof(self.source, self.eof));
        };

        if set.contains(&token.kind()) {
            self.position += 1;
            Ok(*token)
        } else {
            Err(error::ParseError::unexpected_token_set(
                self.source,
                token.span(),
                token.kind(),
                set,
            ))
        }
    }

    #[allow(unused)]
    fn next_is_any<'a>(&self, set: &'a [TokenKind]) -> Option<Token<'file_name, 'source>> {
        self.tokens
            .get(self.position + 1)
            .filter(|token| set.contains(&token.kind()))
            .copied()
    }

    fn nth(&self, n: usize) -> Option<Token<'file_name, 'source>> {
        self.tokens.get(self.position + n).copied()
    }

    fn put_back(&mut self, n: usize) {
        self.position -= n;
    }
}
