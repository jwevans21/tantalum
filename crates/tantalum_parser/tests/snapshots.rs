use insta::assert_ron_snapshot;

use tantalum_lexer::Lexer;
use tantalum_parser::Parser;

macro_rules! snapshot_from_file {
    ($file_name:ident) => {
        #[test]
        fn $file_name() {
            let source = include_str!(concat!("../../examples/", stringify!($file_name), ".ta"));
            let lexer = Lexer::new(concat!(stringify!($file_name), ".ta"), source);

            let mut parser = Parser::new(lexer);

            let ast = parser.parse();

            assert_ron_snapshot!(ast);
        }
    };
    ($($file_name:ident),*) => {
        $(snapshot_from_file! { $file_name })*
    };
}

snapshot_from_file! {
    start,
    hello_world,
    conditionals,
    loops
}
