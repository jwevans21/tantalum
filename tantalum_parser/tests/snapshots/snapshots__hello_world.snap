---
source: tantalum_parser/tests/snapshots.rs
expression: ast
snapshot_kind: text
---
Ok(AST([
  Spanned(
    span: Span(
      start: Location(
        file_name: "hello_world.ta",
        position: 0,
        line: 1,
        column: 1,
      ),
      end: Location(
        file_name: "hello_world.ta",
        position: 46,
        line: 1,
        column: 47,
      ),
    ),
    data: ExternalFunction(ExternalFunction(
      name: Spanned(
        span: Span(
          start: Location(
            file_name: "hello_world.ta",
            position: 10,
            line: 1,
            column: 11,
          ),
          end: Location(
            file_name: "hello_world.ta",
            position: 16,
            line: 1,
            column: 17,
          ),
        ),
        data: "printf",
      ),
      parameters: Spanned(
        span: Span(
          start: Location(
            file_name: "hello_world.ta",
            position: 16,
            line: 1,
            column: 17,
          ),
          end: Location(
            file_name: "hello_world.ta",
            position: 40,
            line: 1,
            column: 41,
          ),
        ),
        data: [
          Spanned(
            span: Span(
              start: Location(
                file_name: "hello_world.ta",
                position: 17,
                line: 1,
                column: 18,
              ),
              end: Location(
                file_name: "hello_world.ta",
                position: 34,
                line: 1,
                column: 35,
              ),
            ),
            data: Named(NamedParameter(
              name: Spanned(
                span: Span(
                  start: Location(
                    file_name: "hello_world.ta",
                    position: 17,
                    line: 1,
                    column: 18,
                  ),
                  end: Location(
                    file_name: "hello_world.ta",
                    position: 23,
                    line: 1,
                    column: 24,
                  ),
                ),
                data: "format",
              ),
              ty: Spanned(
                span: Span(
                  start: Location(
                    file_name: "hello_world.ta",
                    position: 25,
                    line: 1,
                    column: 26,
                  ),
                  end: Location(
                    file_name: "hello_world.ta",
                    position: 34,
                    line: 1,
                    column: 35,
                  ),
                ),
                data: Pointer(PointerType(
                  ty: Spanned(
                    span: Span(
                      start: Location(
                        file_name: "hello_world.ta",
                        position: 26,
                        line: 1,
                        column: 27,
                      ),
                      end: Location(
                        file_name: "hello_world.ta",
                        position: 34,
                        line: 1,
                        column: 35,
                      ),
                    ),
                    data: Const(ConstType(
                      ty: Spanned(
                        span: Span(
                          start: Location(
                            file_name: "hello_world.ta",
                            position: 32,
                            line: 1,
                            column: 33,
                          ),
                          end: Location(
                            file_name: "hello_world.ta",
                            position: 34,
                            line: 1,
                            column: 35,
                          ),
                        ),
                        data: Named(NamedType(
                          name: Spanned(
                            span: Span(
                              start: Location(
                                file_name: "hello_world.ta",
                                position: 32,
                                line: 1,
                                column: 33,
                              ),
                              end: Location(
                                file_name: "hello_world.ta",
                                position: 34,
                                line: 1,
                                column: 35,
                              ),
                            ),
                            data: "u8",
                          ),
                        )),
                      ),
                    )),
                  ),
                )),
              ),
            )),
          ),
          Spanned(
            span: Span(
              start: Location(
                file_name: "hello_world.ta",
                position: 36,
                line: 1,
                column: 37,
              ),
              end: Location(
                file_name: "hello_world.ta",
                position: 39,
                line: 1,
                column: 40,
              ),
            ),
            data: Variadic,
          ),
        ],
      ),
      return_type: Some(Spanned(
        span: Span(
          start: Location(
            file_name: "hello_world.ta",
            position: 42,
            line: 1,
            column: 43,
          ),
          end: Location(
            file_name: "hello_world.ta",
            position: 45,
            line: 1,
            column: 46,
          ),
        ),
        data: Named(NamedType(
          name: Spanned(
            span: Span(
              start: Location(
                file_name: "hello_world.ta",
                position: 42,
                line: 1,
                column: 43,
              ),
              end: Location(
                file_name: "hello_world.ta",
                position: 45,
                line: 1,
                column: 46,
              ),
            ),
            data: "i32",
          ),
        )),
      )),
    )),
  ),
  Spanned(
    span: Span(
      start: Location(
        file_name: "hello_world.ta",
        position: 48,
        line: 3,
        column: 1,
      ),
      end: Location(
        file_name: "hello_world.ta",
        position: 122,
        line: 7,
        column: 2,
      ),
    ),
    data: Function(Function(
      name: Spanned(
        span: Span(
          start: Location(
            file_name: "hello_world.ta",
            position: 51,
            line: 3,
            column: 4,
          ),
          end: Location(
            file_name: "hello_world.ta",
            position: 55,
            line: 3,
            column: 8,
          ),
        ),
        data: "main",
      ),
      parameters: Spanned(
        span: Span(
          start: Location(
            file_name: "hello_world.ta",
            position: 55,
            line: 3,
            column: 8,
          ),
          end: Location(
            file_name: "hello_world.ta",
            position: 57,
            line: 3,
            column: 10,
          ),
        ),
        data: [],
      ),
      return_type: Some(Spanned(
        span: Span(
          start: Location(
            file_name: "hello_world.ta",
            position: 59,
            line: 3,
            column: 12,
          ),
          end: Location(
            file_name: "hello_world.ta",
            position: 62,
            line: 3,
            column: 15,
          ),
        ),
        data: Named(NamedType(
          name: Spanned(
            span: Span(
              start: Location(
                file_name: "hello_world.ta",
                position: 59,
                line: 3,
                column: 12,
              ),
              end: Location(
                file_name: "hello_world.ta",
                position: 62,
                line: 3,
                column: 15,
              ),
            ),
            data: "i32",
          ),
        )),
      )),
      body: Spanned(
        span: Span(
          start: Location(
            file_name: "hello_world.ta",
            position: 63,
            line: 3,
            column: 16,
          ),
          end: Location(
            file_name: "hello_world.ta",
            position: 122,
            line: 7,
            column: 2,
          ),
        ),
        data: Block(Block(
          statements: [
            Spanned(
              span: Span(
                start: Location(
                  file_name: "hello_world.ta",
                  position: 69,
                  line: 4,
                  column: 5,
                ),
                end: Location(
                  file_name: "hello_world.ta",
                  position: 97,
                  line: 4,
                  column: 33,
                ),
              ),
              data: Expression(FunctionCall(FunctionCall(
                function: Spanned(
                  span: Span(
                    start: Location(
                      file_name: "hello_world.ta",
                      position: 69,
                      line: 4,
                      column: 5,
                    ),
                    end: Location(
                      file_name: "hello_world.ta",
                      position: 75,
                      line: 4,
                      column: 11,
                    ),
                  ),
                  data: Variable(Variable(
                    name: Spanned(
                      span: Span(
                        start: Location(
                          file_name: "hello_world.ta",
                          position: 69,
                          line: 4,
                          column: 5,
                        ),
                        end: Location(
                          file_name: "hello_world.ta",
                          position: 75,
                          line: 4,
                          column: 11,
                        ),
                      ),
                      data: "printf",
                    ),
                  )),
                ),
                arguments: [
                  Spanned(
                    span: Span(
                      start: Location(
                        file_name: "hello_world.ta",
                        position: 76,
                        line: 4,
                        column: 12,
                      ),
                      end: Location(
                        file_name: "hello_world.ta",
                        position: 95,
                        line: 4,
                        column: 31,
                      ),
                    ),
                    data: UnaryOperation(UnaryOperation(
                      operator: Spanned(
                        span: Span(
                          start: Location(
                            file_name: "hello_world.ta",
                            position: 93,
                            line: 4,
                            column: 29,
                          ),
                          end: Location(
                            file_name: "hello_world.ta",
                            position: 95,
                            line: 4,
                            column: 31,
                          ),
                        ),
                        data: Ref,
                      ),
                      operand: Spanned(
                        span: Span(
                          start: Location(
                            file_name: "hello_world.ta",
                            position: 76,
                            line: 4,
                            column: 12,
                          ),
                          end: Location(
                            file_name: "hello_world.ta",
                            position: 93,
                            line: 4,
                            column: 29,
                          ),
                        ),
                        data: Literal(Character(Character(
                          value: Spanned(
                            span: Span(
                              start: Location(
                                file_name: "hello_world.ta",
                                position: 76,
                                line: 4,
                                column: 12,
                              ),
                              end: Location(
                                file_name: "hello_world.ta",
                                position: 93,
                                line: 4,
                                column: 29,
                              ),
                            ),
                            data: "\"Hello, World!\\n\"",
                          ),
                        ))),
                      ),
                    )),
                  ),
                ],
              ))),
            ),
            Spanned(
              span: Span(
                start: Location(
                  file_name: "hello_world.ta",
                  position: 103,
                  line: 6,
                  column: 5,
                ),
                end: Location(
                  file_name: "hello_world.ta",
                  position: 120,
                  line: 6,
                  column: 22,
                ),
              ),
              data: Return(Return(
                value: Some(Spanned(
                  span: Span(
                    start: Location(
                      file_name: "hello_world.ta",
                      position: 110,
                      line: 6,
                      column: 12,
                    ),
                    end: Location(
                      file_name: "hello_world.ta",
                      position: 119,
                      line: 6,
                      column: 21,
                    ),
                  ),
                  data: BinaryOperation(BinaryOperation(
                    left: Spanned(
                      span: Span(
                        start: Location(
                          file_name: "hello_world.ta",
                          position: 110,
                          line: 6,
                          column: 12,
                        ),
                        end: Location(
                          file_name: "hello_world.ta",
                          position: 115,
                          line: 6,
                          column: 17,
                        ),
                      ),
                      data: BinaryOperation(BinaryOperation(
                        left: Spanned(
                          span: Span(
                            start: Location(
                              file_name: "hello_world.ta",
                              position: 110,
                              line: 6,
                              column: 12,
                            ),
                            end: Location(
                              file_name: "hello_world.ta",
                              position: 111,
                              line: 6,
                              column: 13,
                            ),
                          ),
                          data: Literal(Integer(Integer(
                            value: Spanned(
                              span: Span(
                                start: Location(
                                  file_name: "hello_world.ta",
                                  position: 110,
                                  line: 6,
                                  column: 12,
                                ),
                                end: Location(
                                  file_name: "hello_world.ta",
                                  position: 111,
                                  line: 6,
                                  column: 13,
                                ),
                              ),
                              data: "3",
                            ),
                            radix: 10,
                          ))),
                        ),
                        operator: Spanned(
                          span: Span(
                            start: Location(
                              file_name: "hello_world.ta",
                              position: 112,
                              line: 6,
                              column: 14,
                            ),
                            end: Location(
                              file_name: "hello_world.ta",
                              position: 113,
                              line: 6,
                              column: 15,
                            ),
                          ),
                          data: Multiplication,
                        ),
                        right: Spanned(
                          span: Span(
                            start: Location(
                              file_name: "hello_world.ta",
                              position: 114,
                              line: 6,
                              column: 16,
                            ),
                            end: Location(
                              file_name: "hello_world.ta",
                              position: 115,
                              line: 6,
                              column: 17,
                            ),
                          ),
                          data: Literal(Integer(Integer(
                            value: Spanned(
                              span: Span(
                                start: Location(
                                  file_name: "hello_world.ta",
                                  position: 114,
                                  line: 6,
                                  column: 16,
                                ),
                                end: Location(
                                  file_name: "hello_world.ta",
                                  position: 115,
                                  line: 6,
                                  column: 17,
                                ),
                              ),
                              data: "7",
                            ),
                            radix: 10,
                          ))),
                        ),
                      )),
                    ),
                    operator: Spanned(
                      span: Span(
                        start: Location(
                          file_name: "hello_world.ta",
                          position: 116,
                          line: 6,
                          column: 18,
                        ),
                        end: Location(
                          file_name: "hello_world.ta",
                          position: 117,
                          line: 6,
                          column: 19,
                        ),
                      ),
                      data: Addition,
                    ),
                    right: Spanned(
                      span: Span(
                        start: Location(
                          file_name: "hello_world.ta",
                          position: 118,
                          line: 6,
                          column: 20,
                        ),
                        end: Location(
                          file_name: "hello_world.ta",
                          position: 119,
                          line: 6,
                          column: 21,
                        ),
                      ),
                      data: Literal(Integer(Integer(
                        value: Spanned(
                          span: Span(
                            start: Location(
                              file_name: "hello_world.ta",
                              position: 118,
                              line: 6,
                              column: 20,
                            ),
                            end: Location(
                              file_name: "hello_world.ta",
                              position: 119,
                              line: 6,
                              column: 21,
                            ),
                          ),
                          data: "2",
                        ),
                        radix: 10,
                      ))),
                    ),
                  )),
                )),
              )),
            ),
          ],
        )),
      ),
    )),
  ),
]))
